agents/ahv/fence_ahv.py

```diff
@@ -0,0 +1,755 @@
+            logging.error("Unknown error %s", err)
+            raise NutanixClientException(f"API call failed: {err}") from err
+
+        if response.status_code not in self.valid_status_codes:
+            logging.error("API call returned status code %s", response.status_code)
+            raise NutanixClientException(f"API call failed: {response}")
+
+        return response
+
+
+class V4NutanixClient(NutanixClient):
+    """
+    Nutanix V4 API client wrapper class. This implements the
+    necessary methods for listing VMs, getting power state
+    of VMs, and setting power state of VMs.
+    """
+    def __init__(self, host=None, username=None, password=None, verify=False):
+        """
+        Init method
+
+        Args:
+            host(str): Host IP address or hostname
+            username(str): username for Prism Central
+            password(str): password for Prism Central account
+            verify(boolean): ssl verify
+
+        Returns:
+            None
+        """
+        self.host = host
+        self.username = username
+        self.password = password
+        self.verify = verify
+        self.base_url = f"https://{self.host}:{PC_PORT}/api"
+        self.vm_url = f"{self.base_url}/vmm/v{V4_VERSION}/ahv/config/vms"
+        self.task_url = f"{self.base_url}/prism/v{V4_VERSION}/config/tasks"
+        super().__init__(username, password)
+
+    def _get_headers(self, vm_uuid=None):
+        """
+        Get headers for HTTP request.
+
+        Args:
+            vm_uuid(str): VM UUID.
+
+        Returns:
+            dict: HTTP headers.
+        """
+        resp = None
+        headers = {'Accept':'application/json',
+                   'Content-Type': 'application/json'}
+
+        if vm_uuid:
+            try:
+                resp = self._get_vm(vm_uuid)
+            except AHVFenceAgentException as err:
+                logging.error("Unable to retrieve etag")
+                raise AHVFenceAgentException from err
+
+            etag_str = resp.headers['Etag']
+            request_id = str(uuid.uuid1())
+            headers['If-Match'] = etag_str
+            headers['Ntnx-Request-Id'] = request_id
+
+        return headers
+
+    def _get_all_vms(self, filter_str=None, limit=None):
+        """
+        Get a list all registered VMs in Nutanix Prism Central cluster
+
+        Args:
+            filter_str(str): filter string.
+            limit(int): Number of VMs to return
+
+        Returns:
+            str: A json formatted output of details of all VMs.
+        """
+        vm_url = self.vm_url
+
+        if filter_str and limit:
+            vm_url = f"{vm_url}?$filter={filter_str}&$limit={limit}"
+        elif filter_str and not limit:
+            vm_url = f"{vm_url}?$filter={filter_str}"
+        elif limit and not filter_str:
+            vm_url = f"{vm_url}?$limit={limit}"
+
+        logging.info("Sending GET request to get VM details, %s", vm_url)
```

Comment:
Nit: update the message which refers to the list of VMs. This debug message is the same as line 229 so when it comes down to debugging it may cause some confusion!

Reply:
I have 2 spaces between classes as per PEP8 style guide

Reply:
I have changed all other occurrences to vm_uuid but this one. Here, there is no vm_uuid being passed since, this method just returns all VMs by applying any filter string that was passed.